add PID control utility to no-OS

buha opened the pull request:
This PID supports:
- setting Kp, Ki, Kd gains
- clipping output to a certain range to match the range of whatever PID is controlling (for example, for an 8-bit PWM, the algorithm would clip the output to a range between 0 and 255).
- hysteresis
- clipping internal integrator to a certain range to avoid overflow

The no_os_pid_control() must be caled in a loop and user is responsible for detecting when the control should stop by not calling it anymore.

This can be done either by characterizing the algorithm for a particular set of configuration parameters and finding a minimum number of iterations that make the output stable or by looking at the output and detecting when it sits within a satisfactory tolerance from the set point.

buha commented on include/no_os_pid.h:
```diff
@@ -0,0 +1,66 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
```
forgot to add header guard, will fix

CiprianRegus commented on util/no_os_pid.c:
```diff
@@ -0,0 +1,145 @@
+	if (!pid)
+		return -EINVAL;
+
+	free(pid);
```
nitpick: set pid to `NULL`

CiprianRegus commented on util/no_os_pid.c:
```diff
@@ -0,0 +1,145 @@
+	return 0;
+}
+
+/* Perform PID control given a controller, a set-point and a process variable.
+ * This has to be called several times, until output stabilizes.
+ * The PID output is returned by the function. */
```
I guess these functions need doxygen style comment headers.

amiclaus commented on util/no_os_pid.c:
```diff
@@ -0,0 +1,145 @@
+	int64_t D; // derivative component
+
+	if (!pid)
+		return 0;
```
do we want to return a success code here?

buha replied:
So basically what this does, if user passes pid = 0 (the only wrong value we can detect, there can be countless others), the control function returns a constant output.

This is acceptable because it's the kind of error that you catch in development.

Although actually it should output not 0 but min(output_clip.low, output_clip.high).

buha replied:
I guess I will change the function to return a status and the output will be a pointer argument...

amiclaus reviewed (commented):
some comments on my side.

buha commented on include/no_os_pid.h:
```diff
@@ -0,0 +1,66 @@
+
+/* Configuration of the PID. */
+struct no_os_pid_config {
+	unsigned int Kp; // proportional gain (micro-units)
```
I agree in general but everywhere in PID theory I found these parameters written with capital K.
Whomever is maintains this code finds it easier to read Kp compared to p_gain or k_p or whatever because it's exactly as in theory...

buha replied:
same goes for SP and PV

buha commented on util/no_os_pid.c:
```diff
@@ -0,0 +1,145 @@
+#include "no_os_print_log.h"
+
+struct no_os_pid {
+	int iacc; // integral accumulator
```
because user should not have access to the members, this is internal/private stuff

buha commented on util/no_os_pid.c:
```diff
@@ -0,0 +1,145 @@
+int no_os_pid_control(struct no_os_pid *pid, int SP, int PV)
+{
+	int ERR; // error
+	int64_t P; // proportional component
```
I've changed ERR, I, P, D  to lowercase in v2

buha commented:
v2:
- added doxygen
- changed some variables to lowercase
- control function outputs status rather than the output (which is now output via a pointer passed as argument)

buha commented on include/no_os_pid.h:
```diff
@@ -0,0 +1,66 @@
+	int low; // low limit of the range
+};
+
+/* Configuration of the PID. */
```
done in v2

buha commented:
V3:
- fixed problem regarding pid control output being returned instead of written to output argument
- moved declaration of variables at the beginning of pid control function
